modules/host/host.go

```diff
@@ -148,6 +158,38 @@ func (h *Host) Capacity() int64 {
 	return h.spaceRemaining
 }
 
+// Close shuts down the host, preparing it for garbage collection.
+func (h *Host) Close() error {
+	// The order in which things are closed has been explicitly chosen to
+	// minimize turbulence in the event of an error.
+
+	// Save the latest host state.
+	err := func() error {
+		h.mu.Lock()
+		defer h.mu.Unlock()
+		return h.save()
+	}()
```

Comment:
Yeah, the biggest thing I'm concerned about is the number of lines that happen under lock. It's very important to me that it's extremely obvious which code is under lock and which code is not. That's why we've got a few conventions including 'exported functions never under lock', 'non-exported functions always under lock', and 'threaded functions not under lock, but must be called in a gothread'.

I also think the 'gather data under lock, branch out of lock' paradigm is very good, but only as long as the scoping is obvious and explicit. I've changed it for this PR